src/player_controller.rs

```diff
@@ -163,7 +164,9 @@ impl PlayerController {
             }
             WindowEvent::Focused(focused) => {
                 logger.log(&format!("Window focused: {:?}", focused));
-                if let Err(err) = window.set_cursor_grab(*focused) {
```

Comment:
I think we want to only set the cursor grab type to be `Confined` when the window is actually focused, and set it to `None` when it isn't. 

I also noticed some weird behaviour with the focusing, not sure if it's caused by a new version of my OS or of winit or if maybe it was always there and I'm just noticing it now, but it seems like the `set_cursor_grab` function fails when setting the grab mode to `Confined` at the same time as clicking into the window, and a small delay seems to fix it (it's a bit hacky but not a big deal for now as long is it works ok). What do you think about replacing the whole block like this:

```rust
WindowEvent::Focused(focused) => {
    if *focused {
        std::thread::sleep(std::time::Duration::from_millis(100));
    }
    logger.log(&format!("Window focused: {:?}", focused));
    if let Err(err) = window.set_cursor_grab(if *focused {
        CursorGrabMode::Confined
    } else {
        CursorGrabMode::None
    }) {
        logger.log(&format!(
            "Couldn't {:?} cursor: {:?}",
            if *focused { "grab" } else { "release" },
            err
        ))
    }
    window.set_cursor_visible(!*focused);
    self.window_focused = *focused;
}
```

I also need to check if it's working on windows too, I'll reboot quick and check.

Reply:
seems fine on widows with or without this change :)

Reply:
Nice! however just the other day I read about a person who had problems with the CursorGrabMode::Confined under Linux X11, he got this error:

thread 'main' panicked at 'failed to set cursor grab mode: Os(OsError { line: 1333, file: "/home/chris/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.27.5/src/platform_impl/linux/x11/window.rs", error: XMisc("Cursor could not be confined: confine location not viewable") })', src/main.rs:34:10

looking at line 1333 in the file the error is from, we can see that it's caused by Xlib returning GrabNotViewable
we can search for "Xlib GrabNotViewable", which returns documentation for XGrabPointer at https://www.x.org/releases/X11R7.6/doc/libX11/specs/libX11/libX11.html, which states that

    If grab_window is not viewable, it fails and returns GrabNotViewable

looking up "xlib viewable window" leads us the same document, where you can find section "Mapping Windows" and documentation for XMapWindow, which says

    Mapping a window that has an unmapped ancestor does not display the window but marks it as eligible for display when the ancestor becomes mapped. Such a window is called unviewable. When all its ancestors are mapped, the window becomes viewable and will be visible on the screen [...]

It's something to keep in mind, but in general under Linux there is such a great fragmentation between the various composers that I am not surprised by this type of error. Testing all the possibilities under Linux could be painful...

Reply:
Fwiw the error I was getting was "Cursor could not be confined: already confined by another client"

Could be interesting to investigate in the future but doesn't seem worth the effort right now